packages/synthetics-sdk-broken-links/src/broken_links.ts

```diff
@@ -82,31 +83,65 @@ export async function runBrokenLinks(
   let browser: Browser;
   try {
     const options = processOptions(inputOptions);
+    let timeLimitTimeout;
+    let timeLimitresolver = () => {};
+    const timeLimitPromise = new Promise<void>((resolve) => {
+      timeLimitresolver = () => {
+        resolve();
+      };
+      const time_used = Date.now() - new Date(startTime).getTime();
+      timeLimitTimeout = setTimeout(
+        timeLimitresolver,
+        total_timeout_millis - time_used
+      );
+    });
+
+    const followed_links: BrokenLinksResultV1_SyntheticLinkResult[] = [];
+
+    const checkLinksPromise = async () => {
+      // create Browser & origin page then navigate to origin_uri, w/ origin
+      // specific settings
+      browser = await puppeteer.launch({ headless: 'new' });
+      const originPage = await openNewPage(browser);
 
-    // create Browser & origin page then navigate to origin_uri, w/ origin
-    // specific settings
-    browser = await puppeteer.launch({ headless: 'new' });
-    const originPage = await openNewPage(browser);
-
-    const followed_links = [await checkOriginLink(originPage, options)];
-    // if orgin link did not pass exit and return the singular link result
-    if (!followed_links[0].link_passed) {
-      return createSyntheticResult(
-        startTime,
-        runtime_metadata,
-        options,
-        followed_links
+      followed_links.push(
+        await checkOriginLink(
+          originPage,
+          options,
+          startTime,
+          total_timeout_millis
+        )
       );
-    }
 
-    // scrape and organize links to check
-    const linksToFollow: LinkIntermediate[] = await scrapeLinks(
-      originPage,
-      options
-    );
+      // if orgin link is not present or if it did not pass: exit and return the
+      // singular link result
+      if (!followed_links[0].link_passed || !followed_links[0].link_passed) {
+        return;
+      }
 
-    // check all links
-    followed_links.push(...(await checkLinks(browser, linksToFollow, options)));
+      // scrape and organize links to check
+      const linksToFollow: LinkIntermediate[] = await scrapeLinks(
+        originPage,
+        options
+      );
+      // check all links
+      followed_links.push(
+        ...(await checkLinks(
+          browser,
+          linksToFollow,
+          options,
+          startTime,
+          total_timeout_millis
+        ))
+      );
+      return;
+    };
+
+    await Promise.race([timeLimitPromise, checkLinksPromise()]);
```

Comment:
What happens if the origin link isnt reached? I'm trying to think through the full behavior of what should happen with these timeouts, and I'm thinking that at the very least, the origin link should be required to pass.